aten/src/ATen/native/mps/operations/Attention.mm

```diff
@@ -0,0 +1,92 @@
+#include <string>
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <iostream>
+#include <optional>
+
+#include <ATen/core/Tensor.h>
+#include <ATen/native/mps/OperationUtils.h>
+#include <ATen/native/transformers/attention.h>
+#include <ATen/native/transformers/sdp_utils_cpp.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/_scaled_dot_product_attention_math_for_mps_native.h>
+#include <ATen/ops/empty_native.h>
+#endif
+
+namespace at::native {
+
+std::tuple<Tensor, Tensor> _scaled_dot_product_attention_math_mps(const Tensor& query,
+                                                                  const Tensor& key,
+                                                                  const Tensor& value,
+                                                                  const std::optional<Tensor>& attn_mask,
+                                                                  double dropout_p,
+                                                                  bool is_causal,
+                                                                  const std::optional<Tensor>& dropout_mask,
+                                                                  std::optional<double> scale) {
+  using namespace mps;
+  struct CachedGraph : public MPSCachedGraph {
+    CachedGraph(MPSGraph* graph) : MPSCachedGraph(graph) {}
+    MPSGraphTensor* qTensor = nil;
+    MPSGraphTensor* kTensor = nil;
+    MPSGraphTensor* vTensor = nil;
+    MPSGraphTensor* attnTensor = nil;
+    MPSGraphTensor* outputTensor = nil;
+  };
+  int64_t batchSize = query.size(0);
+  int64_t qSize = query.size(2);
+  int64_t num_head = query.size(1);
+  int64_t headSize = query.size(3);
+  auto out = at::empty({batchSize, num_head, qSize, headSize}, query.options());
+  auto scale_factor = sdp::calculate_scale(query, scale).sqrt().as_float_unchecked();
```

Comment:
fyi, there's a native op in MPSGraph starting with Sequoia https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/scaleddotproductattentionwithquerytensor:keytensor:valuetensor:scale:name:  (masked one: https://developer.apple.com/documentation/metalperformanceshadersgraph/mpsgraph/scaleddotproductattentionwithquerytensor:keytensor:valuetensor:masktensor:scale:name:)

Reply:
says the page cannot be found?

BTW is the native op implementation of flash attention?